services/galley/test/integration/API/MLS/Util.hs

```diff
@@ -316,12 +319,54 @@ takeLastPrekeyNG = do
       pure pk
     [] -> error "no prekeys left"
 
+toRandomFile :: ByteString -> MLSTest FilePath
+toRandomFile bs = do
+  p <- randomFileName
+  liftIO $ BS.writeFile p bs
+  pure p
+
+randomFileName :: MLSTest FilePath
+randomFileName = do
+  bd <- State.gets mlsBaseDir
+  (bd </>) . UUID.toString <$> liftIO UUIDV4.nextRandom
+
 mlscli :: HasCallStack => ClientIdentity -> [String] -> Maybe ByteString -> MLSTest ByteString
 mlscli qcid args mbstdin = do
   bd <- State.gets mlsBaseDir
   let cdir = bd </> cid2Str qcid
-  liftIO $ do
-    spawn (proc "mls-test-cli" (["--store", cdir </> "store"] <> args)) mbstdin
+
+  groupOut <- randomFileName
+  let substOut = argSubst "<group-out>" groupOut
+
+  hasState <- hasClientGroupState qcid
+  substIn <-
+    if hasState
+      then do
+        gs <- getClientGroupState qcid
+        fn <- toRandomFile gs
+        pure (argSubst "<group-in>" fn)
+      else pure mempty
+
+  out <-
+    liftIO $
+      spawn
+        ( proc
+            "mls-test-cli"
+            ( ["--store", cdir </> "store"]
+                <> map (appEndo (substIn <> substOut)) args
```

Comment:
It seems to me this is the only place where the monoid instance of `Endo` is used. So why not simply replace with a function and use `.` here?

Reply:
I think I got a little bit to excited when I realized that I have an endomorphism, it's really not needed. I'm going to replace it with plain functions and use the IsString strategy as well